### PR TITLE
Flyer test

### PR DESCRIPTION
**Pull request by JunAishima**
Creating objects for adding the flyer object used for standard collection (zebra + Galil vector + Pilatus) to LSDC. Pull requests and fixes while testing on the beamline.

### CODE REVIEW

**Comment by RobertSchaffer1** [comment]
This'll be a big one, I see.

**Comment by JunAishima** [comment]
> This'll be a big one, I see.

l think this is a lot bigger than necessary - a lot of the code has already been reviewed, but we haven't merged into main yet. Let me see if I can try to clean things up a bit...

File: nyxtools/_version.py [hunk 1]
```diff
@@ -190,27 +196,34 @@ def git_versions_from_keywords(keywords, tag_prefix, verbose):
...
+        tags = set([r for r in refs if re.search(r"\d", r)])
         if verbose:
             print("discarding '%s', no digits" % ",".join(refs - tags))
     if verbose:
         print("likely tags: %s" % ",".join(sorted(tags)))
     for ref in sorted(tags):
         # sorting will prefer e.g. "2.0" over "2.0rc1"
         if ref.startswith(tag_prefix):
-            r = ref[len(tag_prefix):]
+            r = ref[len(tag_prefix) :]
```
Comment (RobertSchaffer1) [inline nyxtools/_version.py:207]: As much as I hate this, I suppose it must be suggested by black/isort/flake?

**Comment by RobertSchaffer1** [comment]
> > This'll be a big one, I see.
> 
> l think this is a lot bigger than necessary - a lot of the code has already been reviewed, but we haven't merged into main yet. Let me see if I can try to clean things up a bit...

Ah, okay I'll wait for that then.

File: nyxtools/_version.py (hunk 1 above)
Comment (mrakitin) [reply nyxtools/_version.py]: I did not see it in my latest commit https://github.com/NSLS-II-NYX/nyxtools/pull/10/commits/b96745f60da70ad23ef71d1a422a38a002555269, but probably was addressed by the earlier linting. And we usually use this `_version.py` file as is coming from https://github.com/NSLS-II/scientific-python-cookiecutter, so I would not worry about it too much. In some projects, we disable checking against `versioneer.py` and this file in the library.

**Comment by JunAishima** [comment]
UPDATED! I rebased on main after merging main with the last pull request ec6eb3d. This should contain only the changes done in the last 10 days or so and that have not been reviewed yet.

File: nyxtools/flyer.py [hunk 2]
```diff
@@ -20,93 +23,183 @@ def __init__(self, vector, zebra, detector=None) -> None:
...
+
+        super().update_parameters(**kwargs)
+
     def kickoff(self):
         self.detector.stage()
-        self.vector.move()
 
-        return NullStatus()
+        def zebra_callback(*args, **kwargs):
+            print(f"\n{ttime.ctime()}:\n\targs: {args}\n\tkwargs: {kwargs}\n")
```
Comment (RobertSchaffer1) [inline nyxtools/flyer.py:56]: We should replace these with logger.info or logger.debug statements, if we want to leave them in.

Comment (RobertSchaffer1) [reply nyxtools/flyer.py:56]: I can do that, if you guys agree.

File: nyxtools/flyer.py [hunk 3]
```diff
@@ -20,93 +23,183 @@ def __init__(self, vector, zebra, detector=None) -> None:
...
+
+        return st_vector & st_detector
+
+    def describe_collect(self):
+        return_dict = {
+            "primary": {
+                f"{self.detector.name}_image": {
+                    "source": f"{self.detector.name}_image",
+                    "dtype": "array",
+                    "shape": [-1, -1],  # TODO: read shapes from AD
```
Comment (RobertSchaffer1) [inline nyxtools/flyer.py:91]: I can do this, unless you're already on it @mrakitin

File: nyxtools/flyer.py [hunk 4]
```diff
@@ -20,93 +23,183 @@ def __init__(self, vector, zebra, detector=None) -> None:
...
-            "data": data,
-            "timestamps": {key: now for key in data},
-            "time": now,
-            "filled": {key: False for key in data},
-        }
+        for datum_id in self._datum_ids:
+            now = ttime.time()
+            data = {
+                # f"{self.detector.name}_image": self._datum_ids["data"],
+                # "omega": self._datum_ids["omega"],
```
Comment (RobertSchaffer1) [inline nyxtools/flyer.py:105]: Do we want to get this back in?

File: nyxtools/vector.py [hunk 5]
```diff
@@ -232,28 +233,50 @@ def prepare_move(
...
                 return True
             else:
+                print(f"{time.ctime()}: changing {old_value} -> {value}...")
                 return False
 
-        run_status = SubscriptionStatus(self.running, start_callback, run=False)
+        run_status = SubscriptionStatus(self.state, start_callback, run=True)
+        print(f"{time.ctime()}: subscribed to {self.active.name}")
+
+        time.sleep(1.0)
```
Comment (mrakitin) [inline nyxtools/vector.py:264]: Why did we need this 1s timeout here?

**Review by mrakitin — APPROVED** [review]
I think it's in a good shape now. I posted a few items to address on Friday.

File: nyxtools/vector.py [hunk 6]
```diff
@@ -232,28 +233,50 @@ def prepare_move(
...
+        #          'enum_strs': ('No', 'Yes'),
+        #          'sub_type': 'value',
+        #          'obj': EpicsSignalRO(read_pv='XF:19IDC-ES{Gon:1-Vec}Sts:Running-Sts',
+        #                               name='vector_active', parent='vector', value=0,
+        #                               timestamp=1638904545.824989, auto_monitor=False,
+        #                               string=False)}
+        def start_callback(value, old_value, **kwargs):
+            print(f"{time.ctime()}: {old_value} -> {value}")
+            if old_value == "Backup" and value == "Acquiring":
+                print(f"{time.ctime()}: Successfully changed {old_value} -> {value}")
```
Comment (RobertSchaffer1) [inline nyxtools/vector.py:255]: While we're changing these to debugs we should remember to change this to ttime as well.

File: nyxtools/flyer.py (hunk 3 above)
Comment (mrakitin) [reply nyxtools/flyer.py:91]: Sounds great! Please have a look at https://github.com/bluesky/databroker/pull/596/files#diff-f7833108df15eaaac29030bc84ed7b307eb03b4eb91d959b4917d94ee1d13004R2097-R2101. That link has a few comments about what shape we expect (`(num_images, y, x)`).

File: nyxtools/flyer.py (hunk 2 above)
Comment (mrakitin) [reply nyxtools/flyer.py:56]: Yes, please go ahead, @RobertSchaffer1. I'll be happy to review your changes.

**Review by RobertSchaffer1 — APPROVED** [review]
Agreed w/ Maksim, most of this is fine. Could use some basic cleanup.

File: nyxtools/vector.py (hunk 6 above)
Comment (mrakitin) [reply nyxtools/vector.py:255]: Right, good point.

File: nyxtools/vector.py (hunk 5 above)
Comment (RobertSchaffer1) [reply nyxtools/vector.py:264]: I can't think of a single good reason, which makes me think it's an artifact of the original code. I'll remove it.

**Comment by RobertSchaffer1** [comment]
@JunAishima @mrakitin I made some minor changes, just the ones we discussed on this PR.

**Review by mrakitin — APPROVED** [review]
Looks good! Had a discussion with @RobertSchaffer1 about the log message format, and how to enable the `DEBUG` stdout logging on the client-side. This config may be useful:
https://github.com/NSLS-II/sirepo-bluesky/blob/0aed9b9c735c94728ce360345f09b6c4722bb583/sirepo_bluesky/sirepo_ophyd.py#L20-L24

**Pull request merged**